File: cses/09_geometry/3430_lines-and-queries-ii/solution.py
import sys


def main():
    data = sys.stdin.buffer.read().split()
    pos = 0
    n = int(data[pos]); pos += 1

    # Li Chao tree đặt trên trục vị trí, phủ [0, SZ) với SZ là lũy thừa 2 >= M+1.
    M = 100000
    SZ = 1
    while SZ < M + 1:
        SZ <<= 1                 # SZ = 131072 = 2^17
    H = SZ.bit_length() - 1      # chiều cao cây = 17

    LA = [None] * (2 * SZ)       # hệ số góc (slope) lưu tại mỗi nút (None = rỗng)
    LB = [None] * (2 * SZ)       # hệ số tự do (intercept) lưu tại mỗi nút

    out = []
    b1 = b'1'

    for _ in range(n):
        if data[pos] == b1:
            # Loại 1: thêm đường y = a*x + b active trên đoạn vị trí [l, r].
            a = int(data[pos + 1]); b = int(data[pos + 2])
            l = int(data[pos + 3]); r = int(data[pos + 4])
            pos += 5

            # Phân rã canonical [l, r] (nửa mở [l, r+1)) thành O(log M) nút chuẩn.
            lp = l + SZ
            rp = r + SZ + 1
            nodes = []
            while lp < rp:
                if lp & 1:
                    nodes.append(lp); lp += 1
                if rp & 1:
                    rp -= 1; nodes.append(rp)
                lp >>= 1; rp >>= 1

            for node0 in nodes:
                # Đoạn vị trí [lo, hi] mà nút chuẩn node0 phủ.
                L = node0.bit_length() - 1
                span = SZ >> L
                lo = (node0 - (1 << L)) * span
                hi = lo + span - 1
                # Li Chao insert đường (a, b) vào subtree gốc là node0.
                node = node0
                m = a; k = b
                while True:
                    cm = LA[node]
                    if cm is None:
                        # Nút trống: đặt luôn đường vào đây.
                        LA[node] = m; LB[node] = k
                        break
                    ck = LB[node]
                    mid = (lo + hi) >> 1
                    # So sánh tại điểm giữa mid: giữ đường tốt hơn ở nút.
                    if m * mid + k > cm * mid + ck:
                        LA[node] = m; LB[node] = k
                        wm = m; wk = k
                        m = cm; k = ck          # đường thua = đường cũ ở nút
                    else:
                        wm = cm; wk = ck        # đường thua = đường mới đưa vào
                    if lo == hi:
                        break
                    # Đẩy đường thua xuống nửa mà nó còn có thể vượt (xét đầu mút trái).
                    if m * lo + k > wm * lo + wk:
                        node = node * 2; hi = mid
                    else:
                        node = node * 2 + 1; lo = mid + 1
        else:
            # Loại 2: đi root -> leaf(x), lấy max các đường lưu trên đường đi.
            x = int(data[pos + 1]); pos += 2
            leaf = SZ + x
            best = None
            for s in range(H, -1, -1):
                node = leaf >> s
                a = LA[node]
                if a is not None:
                    v = a * x + LB[node]
                    if best is None or v > best:
                        best = v
            out.append(b"NO" if best is None else str(best).encode())

    sys.stdout.buffer.write(b"\n".join(out))
    if out:
        sys.stdout.buffer.write(b"\n")


main()

File: cses/09_geometry/3430_lines-and-queries-ii/solution.cpp
#include <algorithm>
#include <iostream>
#include <utility>
#include <vector>

// Một đường thẳng y = slope*x + intercept.
struct Line {
    long long slope = 0;
    long long intercept = 0;

    long long value(const int x) const {
        return slope * x + intercept;
    }
};

// Li Chao tree đặt trên TRỤC VỊ TRÍ [0, maximum_x]; mỗi nút giữ tối đa một đường.
class SegmentLiChaoTree {
public:
    explicit SegmentLiChaoTree(const int maximum_x)
        : maximum_x_(maximum_x),
          lines_(static_cast<std::size_t>(4 * (maximum_x + 1))),
          occupied_(static_cast<std::size_t>(4 * (maximum_x + 1)), false) {}

    // Thêm đường active trên đoạn vị trí [query_left, query_right].
    void insert(const Line line, const int query_left, const int query_right) {
        insert_segment(
            line, query_left, query_right, 1, 0, maximum_x_
        );
    }

    // Truy vấn max của mọi đường active tại vị trí x.
    std::pair<bool, long long> query(const int x) const {
        return query(x, 1, 0, maximum_x_);
    }

private:
    // Li Chao insert: chèn đường vào subtree gốc node phủ đoạn [left, right].
    void insert_line(Line incoming, const int node, const int left,
                     const int right) {
        // Nút trống: đặt luôn đường vào đây.
        if (!occupied_[static_cast<std::size_t>(node)]) {
            occupied_[static_cast<std::size_t>(node)] = true;
            lines_[static_cast<std::size_t>(node)] = incoming;
            return;
        }

        // So sánh tại điểm giữa: giữ đường tốt hơn ở nút, chuẩn bị đẩy đường thua.
        const int middle = (left + right) / 2;
        const bool better_left =
            incoming.value(left) > lines_[static_cast<std::size_t>(node)].value(left);
        const bool better_middle =
            incoming.value(middle) > lines_[static_cast<std::size_t>(node)].value(middle);
        if (better_middle) {
            std::swap(incoming, lines_[static_cast<std::size_t>(node)]);
        }
        if (left == right) {
            return;
        }
        // Đẩy đường thua xuống nửa mà nó còn có thể vượt (quyết định ở đầu mút trái).
        if (better_left != better_middle) {
            insert_line(incoming, 2 * node, left, middle);
        } else {
            insert_line(incoming, 2 * node + 1, middle + 1, right);
        }
    }

    // Phân rã [query_left, query_right] thành các nút chuẩn rồi Li Chao insert từng nút.
    void insert_segment(const Line line, const int query_left,
                        const int query_right, const int node,
                        const int left, const int right) {
        if (query_right < left || right < query_left) {
            return;
        }
        // Nút nằm gọn trong [l, r]: đây là nút chuẩn, chèn cả đường vào subtree của nó.
        if (query_left <= left && right <= query_right) {
            insert_line(line, node, left, right);
            return;
        }
        const int middle = (left + right) / 2;
        insert_segment(
            line, query_left, query_right, 2 * node, left, middle
        );
        insert_segment(
            line, query_left, query_right, 2 * node + 1, middle + 1, right
        );
    }

    // Đi từ gốc xuống lá x, lấy max giá trị các đường lưu trên đường đi root -> leaf.
    std::pair<bool, long long> query(const int x, const int node,
                                     const int left, const int right) const {
        bool found = occupied_[static_cast<std::size_t>(node)];
        long long best = found
            ? lines_[static_cast<std::size_t>(node)].value(x)
            : 0;
        if (left == right) {
            return {found, best};
        }

        const int middle = (left + right) / 2;
        const auto child = x <= middle
            ? query(x, 2 * node, left, middle)
            : query(x, 2 * node + 1, middle + 1, right);
        if (!found) {
            return child;
        }
        if (!child.first) {
            return {true, best};
        }
        return {true, std::max(best, child.second)};
    }

    int maximum_x_;
    std::vector<Line> lines_;
    std::vector<char> occupied_;
};

int main() {
    std::ios::sync_with_stdio(false);
    std::cin.tie(nullptr);

    constexpr int MAXIMUM_X = 100'000;
    int query_count;
    std::cin >> query_count;
    SegmentLiChaoTree tree(MAXIMUM_X);
    while (query_count-- > 0) {
        int type;
        std::cin >> type;
        if (type == 1) {
            // Loại 1: thêm đường y = a*x + b active trên [l, r].
            Line line;
            int left, right;
            std::cin >> line.slope >> line.intercept >> left >> right;
            tree.insert(line, left, right);
        } else {
            // Loại 2: in max tại x, hoặc "NO" nếu không có đường active.
            int x;
            std::cin >> x;
            const auto answer = tree.query(x);
            if (answer.first) {
                std::cout << answer.second << '\n';
            } else {
                std::cout << "NO\n";
            }
        }
    }
    return 0;
}
